src/pactServer.ts

```diff
@@ -0,0 +1,295 @@
+const SERVER_STARTUP_SUCCESS_LINES = [
+  'No replay found',
+  'starting on port',
+]
+
+// Log lines that indicate a failure on start up.
+const SERVER_STARTUP_FAILURE_LINES = [
+  'Address already in use',
+]
+
+type LogCallback = (line: string) => void | Promise<void>
+
+export interface PactServerOptions {
+  /**
+   * Called on lines sent to stdout or stderr for SERVER_STARTUP_TIMEOUT_MS after calling start().
+   */
+  onStartupLogLine?: LogCallback,
+
+  /**
+   * Called on every stdout line, including during startup.
+   */
+  onStdoutLine?: LogCallback,
+
+  /**
+   * Called on every stderr line, including during startup.
+   */
+  onStderrLine?: LogCallback,
+
+  /**
+   * If false, will apply default loggers for stdout and stderr.
+   */
+  quiet?: boolean,
+}
+
+/**
+ * Wrapper for using Node child processes and handling output.
+ */
+class Process {
+  private readonly path: string
+  private readonly args: string[]
+  private readonly options: ExecOptions
+  private readonly stdoutCallbacks: LogCallback[] = []
+  private readonly stderrCallbacks: LogCallback[] = []
+  private childProcess?: ChildProcess
+  private _onExitPromise?: Promise<void>
+  private _started = false
+
+  constructor(path: string, args: string[], options: ExecOptions = {}) {
+    this.path = path
+    this.args = args
+    this.options = _.defaults({
+      stdio: ['ignore', 'pipe', 'pipe'],
+    }, options)
+  }
+
+  get started(): boolean {
+    return this._started
+  }
+
+  get onExitPromise(): Promise<void> {
+    if (!this._started) {
+      throw new Error('Process not yet started.')
+    }
+    return this._onExitPromise!
+  }
+
+  start(): void {
+    if (this._started) {
+      throw new Error('Process already started.')
+    }
+    this._started = true
+    this.childProcess = spawn(this.path, this.args, this.options)
+
+    setImmediate(async () => {
+      for await (const line of chunksToLinesAsync(this.childProcess!.stdout as AsyncIterable<string>)) {
+        await Promise.all(_.map(this.stdoutCallbacks, async (callback: LogCallback) => { await callback(chomp(line)) }))
+      }
+    })
+
+    setImmediate(async () => {
+      for await (const line of chunksToLinesAsync(this.childProcess!.stderr as AsyncIterable<string>)) {
+        await Promise.all(_.map(this.stderrCallbacks, async (callback: LogCallback) => { await callback(chomp(line)) }))
+      }
+    })
+
+    this._onExitPromise = new Promise((resolve, reject) => {
+      this.childProcess!.once('exit', (code: number, _signal: string) => {
+        if (code === 0) {
+          resolve()
+        } else {
+          reject(new Error(`Process exited with error code: ${code}`))
+        }
+      })
+      this.childProcess!.once('error', (error: Error) => {
+        reject(error)
+      })
+    })
+  }
+
+  kill(): void {
+    if (!this.childProcess) {
+      throw new Error('Process was not started.')
+    }
+    this.childProcess.kill()
```

Comment:
do we no longer need the kill by PID hack? this works in all scenarios? or is it that it works now because we also added the `process.on('exit'` code